framework/src/messaging/messaging_send.c

```diff
@@ -0,0 +1,186 @@
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <mqueue.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <messaging/messaging.h>
+#include "messaging_internal.h"
+
+#define MSG_REPLY_PRIO 10
+/****************************************************************************
+ * private functions
+ ****************************************************************************/
+#ifdef CONFIG_MESSAGING_ENABLE_HEADER
+static int messaging_append_header(const char *msg, int msglen, messaging_header_t *data)
+{
+	data->sender_pid = getpid();
+	data->msglen = msglen;
+	data->msg = (char *)MSG_ALLOC(msglen);
+	if (data->msg == NULL) {
+		return ERROR;
+	}
+	memcpy(data->msg, msg, msglen);
+	return OK;
+}
+#endif
+
+static int messaging_sync_recv(const char *port_name, msg_send_param_t *param)
+{
+	int ret = OK;
+	mqd_t sync_mqdes;
+	char *sync_portname;
+	struct mq_attr internal_attr;
+
+	internal_attr.mq_maxmsg = CONFIG_MESSAGING_UNICAST_MAXMSG;
+	internal_attr.mq_msgsize = CONFIG_MESSAGING_UNICAST_MSGSIZE;
+	internal_attr.mq_flags = 0;
+
+	/* sender waits the reply with "port_name + MSG_SYNC_PORTNAME_PREFIX". */
+	MSG_ASPRINTF(&sync_portname, "%s%s", port_name, MSG_SYNC_PORTNAME_PREFIX);
+	if (sync_portname == NULL) {
+		msgdbg("message send fail : sync portname allocation fail.\n");
+		return ERROR;
+	}
+	sync_mqdes = mq_open(sync_portname, O_RDONLY | O_CREAT, 0666, &internal_attr);
+	if (sync_mqdes == (mqd_t)ERROR) {
+		msgdbg("message send fail : sync open fail %d.\n", errno);
+		MSG_FREE(sync_portname);
+		return ERROR;
+	}
+
+	ret = mq_receive(sync_mqdes, (char *)param->sync_msg, param->sync_msglen, 0);
+	if (ret != OK) {
+		msgdbg("message send fail : sync recv fail %d.\n", errno);
+		ret = ERROR;
+	}
+
+	MSG_FREE(sync_portname);
+	mq_close(sync_mqdes);
+	mq_unlink(sync_portname);
+
+	return ret;
+}
+
+static int messaging_send_internal(mqd_t *mqdes, const char *port_name, const char *msg, size_t msglen, int prio, void *data)
+{
+	int ret;
+	struct mq_attr internal_attr;
+
+	internal_attr.mq_maxmsg = CONFIG_MESSAGING_UNICAST_MAXMSG;
+	internal_attr.mq_msgsize = CONFIG_MESSAGING_UNICAST_MSGSIZE;
+	internal_attr.mq_flags = 0;
+
+	*mqdes = mq_open(port_name, O_WRONLY | O_CREAT, 0666, &internal_attr);
+	if (*mqdes == (mqd_t)ERROR) {
+		msgdbg("message send fail : open fail %d.\n", errno);
+		return ERROR;
+	}
+
+#ifdef CONFIG_MESSAGING_ENABLE_HEADER
+	ret = mq_send(*mqdes, (char *)data, sizeof(messaging_header_t) + msglen, prio);
+#else
+	ret = mq_send(*mqdes, msg, msglen, prio);
+#endif
+	if (ret != OK) {
+		msgdbg("message send fail : send fail %d.\n", errno);
```

Comment:
need mq_close and mq_unlink

Reply:
1. close the mq in messaging_send(). but I think moving to here is better. I will fix.
2. mq_unlink is not needed. Because if unlink the queue, receiver cannot receive the message.